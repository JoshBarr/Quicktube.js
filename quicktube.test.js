// Mock for Google Analytics.
const ga = jest.fn();

// Mocks for YouTube iframe API.
let playerEvents;
const mockPlayer = {
    playVideo: jest.fn(),
    pauseVideo: jest.fn(),
};

const YT = {
    PlayerState: {
        BUFFERING: 3,
        CUED: 5,
        ENDED: 0,
        PAUSED: 2,
        PLAYING: 1,
        UNSTARTED: -1,
    },
    // Mock the YT.Player API, return a mock player object and store events.
    Player: jest.fn((playerId, config) => {
        playerEvents = config.events;
        return mockPlayer;
    }),
};

// Mock for
window.ga = ga;
window.YT = YT;

const mockHTML = `
<script></script>
<div class="quicktube" data-quicktube="k6QanQUaDOo">
    <div class="quicktube__video quicktube__video--paused" data-quicktube-video></div>
    <div data-quicktube-play="k6QanQUaDOo" class="quicktube__poster" data-quicktube-poster>
        <div class="quicktube__play quicktube__btn" tabindex="0">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop="k6QanQUaDOo">Stop button</div>
`;

const simulateEvent = (selector, type, data = {}) => {
    const el = typeof selector === 'string' ? document.querySelector(selector) : selector;
    const event = document.createEvent('Events');
    event.initEvent(type, true, false);

    el.dispatchEvent(Object.assign(event, data));
};

describe('Quicktube', () => {
    document.body.innerHTML = mockHTML;
    const Quicktube = require('./quicktube');
    const loadedHTML = document.body.innerHTML;

    it('exists', () => {
        expect(Quicktube).toBeDefined();
    });

    it('loaded', () => {
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    describe('onYouTubeIframeAPIReady', () => {
        it('exists', () => {
            expect(window.onYouTubeIframeAPIReady).toBeDefined();
        });

        it('initialises on page ready', () => {
            document.body.innerHTML = loadedHTML;
            window.onYouTubeIframeAPIReady();
            simulateEvent(document, 'DOMContentLoaded');
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('init', () => {
        it('default options', () => {
            Quicktube.init();
            expect(Quicktube.options).toEqual({
                trackAnalytics: false,
                activeClass: 'quicktube--playing',
                pausedClass: 'quicktube--paused',
                posterFrameHiddenClass: 'quicktube__poster--hidden',
                showInfo: 0,
                autohide: 1,
                color: 'white',
                wmode: 'transparent',
            });
        });

        it('custom options', () => {
            Quicktube.init({ trackAnalytics: true });
            expect(Quicktube.options).toEqual({
                trackAnalytics: true,
                activeClass: 'quicktube--playing',
                pausedClass: 'quicktube--paused',
                posterFrameHiddenClass: 'quicktube__poster--hidden',
                showInfo: 0,
                autohide: 1,
                color: 'white',
                wmode: 'transparent',
            });
        });
    });

    describe('play', () => {
        beforeEach(() => {
            document.body.innerHTML = loadedHTML;
            // const quicktube = Quicktube.init({ trackAnalytics: true });
        });

        it('click', () => {
            simulateEvent('[data-quicktube-play]', 'click');
            expect(document.body.innerHTML).toMatchSnapshot();
            expect(Quicktube.quicktubePlayer).toBe(mockPlayer);
        });

        it('keydown wrong key', () => {
            simulateEvent('[data-quicktube-play]', 'keydown', { keyCode: 50 });
            expect(document.body.innerHTML).toMatchSnapshot();
        });

        it('keydown right key', () => {
            simulateEvent('[data-quicktube-play]', 'keydown', { keyCode: 13 });
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('stop', () => {
        beforeEach(() => {
            document.body.innerHTML = loadedHTML;
            // const quicktube = Quicktube.init({ trackAnalytics: true });
        });

        it('click', () => {
            simulateEvent('[data-quicktube-stop]', 'click');
            expect(document.body.innerHTML).toMatchSnapshot();
            expect(Quicktube.quicktubePlayer).toBe(mockPlayer);
        });
    });

    describe.skip('play iOS', () => {
        beforeEach(() => {
            document.body.innerHTML = loadedHTML;
            navigator.userAgent = 'Mozilla/5.0 (iPhone; CPU iPhone OS 9_1 like Mac OS X) AppleWebKit/601.1.46 (KHTML, like Gecko) Version/9.0 Mobile/13B143 Safari/601.1';
            Quicktube.init();
        });

        it('click', () => {
            simulateEvent('[data-quicktube-play]', 'click');
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('YT Player events', () => {
        beforeEach(() => {
            document.body.innerHTML = loadedHTML;
            // const quicktube = Quicktube.init({ trackAnalytics: true });
            simulateEvent('[data-quicktube-play]', 'click');
            ga.mockClear();
            mockPlayer.pauseVideo.mockClear();
        });

        describe('onReady', () => {
            it('not playing - resume video', () => {
                const playVideo = jest.fn();
                playerEvents.onReady({ target: { playVideo } });
                expect(playVideo).toHaveBeenCalled();
            });

            it('playing - stop video', () => {
                const playVideo = jest.fn();
                playerEvents.onReady({ target: { playVideo } });
                expect(playVideo).toHaveBeenCalled();
                Quicktube.stopVideo('kittens');
                expect(mockPlayer.pauseVideo).toHaveBeenCalled();
            });
        });

        describe('onStateChange', () => {
            it('PLAYING', () => {
                playerEvents.onStateChange({
                    target: {
                        getVideoData: jest.fn(() => ({ title: 'test title' })),
                    },
                    data: YT.PlayerState.PLAYING,
                });
                expect(document.body.innerHTML).toMatchSnapshot();
                expect(ga.mock.calls[0]).toMatchSnapshot();
            });

            describe('PLAYING monitoring', () => {
                it('still PLAYING 0%', () => {
                    jest.useFakeTimers();
                    playerEvents.onStateChange({
                        target: {
                            getVideoData: jest.fn(() => ({ title: 'test title' })),
                            getPlayerState: jest.fn(() => YT.PlayerState.PLAYING),
                            getDuration: jest.fn(() => 100),
                            getCurrentTime: jest.fn(() => 0),
                            // lastP: '',
                        },
                        data: YT.PlayerState.PLAYING,
                    });
                    jest.runOnlyPendingTimers();
                    expect(ga.mock.calls[1]).toMatchSnapshot();
                });

                it('still PLAYING 25%', () => {
                    jest.useFakeTimers();
                    playerEvents.onStateChange({
                        target: {
                            getVideoData: jest.fn(() => ({ title: 'test title' })),
                            getPlayerState: jest.fn(() => YT.PlayerState.PLAYING),
                            getDuration: jest.fn(() => 100),
                            getCurrentTime: jest.fn(() => 25),
                            // lastP: '',
                        },
                        data: YT.PlayerState.PLAYING,
                    });
                    jest.runOnlyPendingTimers();
                    expect(ga.mock.calls[1]).toMatchSnapshot();
                });

                it('still PLAYING 50%', () => {
                    jest.useFakeTimers();
                    playerEvents.onStateChange({
                        target: {
                            getVideoData: jest.fn(() => ({ title: 'test title' })),
                            getPlayerState: jest.fn(() => YT.PlayerState.PLAYING),
                            getDuration: jest.fn(() => 100),
                            getCurrentTime: jest.fn(() => 50),
                            // lastP: '',
                        },
                        data: YT.PlayerState.PLAYING,
                    });
                    jest.runOnlyPendingTimers();
                    expect(ga.mock.calls[1]).toMatchSnapshot();
                });

                it('still PLAYING 100%', () => {
                    jest.useFakeTimers();
                    playerEvents.onStateChange({
                        target: {
                            getVideoData: jest.fn(() => ({ title: 'test title' })),
                            getPlayerState: jest.fn(() => YT.PlayerState.PLAYING),
                            getDuration: jest.fn(() => 100),
                            getCurrentTime: jest.fn(() => 50),
                            // lastP: '',
                        },
                        data: YT.PlayerState.PLAYING,
                    });
                    jest.runOnlyPendingTimers();
                    expect(ga.mock.calls[1]).toMatchSnapshot();
                });
            });

            it('PAUSED', () => {
                playerEvents.onStateChange({
                    target: {
                        getVideoData: jest.fn(() => ({ title: 'test title' })),
                    },
                    data: YT.PlayerState.PAUSED,
                });
                expect(document.body.innerHTML).toMatchSnapshot();
                expect(ga.mock.calls[0]).toMatchSnapshot();
            });

            it('ENDED', () => {
                playerEvents.onStateChange({
                    target: {
                        getVideoData: jest.fn(() => ({ title: 'test title' })),
                    },
                    data: YT.PlayerState.ENDED,
                });
                expect(document.body.innerHTML).toMatchSnapshot();
                expect(mockPlayer.pauseVideo).toHaveBeenCalled();
            });
        });

        it('onError', () => {
            playerEvents.onError({ target: { src: 'test src' }, data: 'test data' });
            expect(ga.mock.calls[0]).toMatchSnapshot();
        });
    });
});
