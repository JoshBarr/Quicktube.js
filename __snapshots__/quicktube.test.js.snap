// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quicktube YT Player events onError 1`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "GTM",
  "youtube:test src-test data",
]
`;

exports[`Quicktube YT Player events onStateChange ENDED 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube quicktube--paused\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    <iframe src=\\"https://www.youtube.com/embed/k6QanQUaDOo?autoplay=1&amp;showinfo=0&amp;autohide=1&amp;color=white&amp;enablejsapi=1&amp;playerapiid=ytplayer&amp;wmode=transparent\\" width=\\"100%\\" id=\\"kittens\\" class=\\"quicktube__iframe\\"></iframe></div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube YT Player events onStateChange PAUSED 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube quicktube--playing\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    <iframe src=\\"https://www.youtube.com/embed/k6QanQUaDOo?autoplay=1&amp;showinfo=0&amp;autohide=1&amp;color=white&amp;enablejsapi=1&amp;playerapiid=ytplayer&amp;wmode=transparent\\" width=\\"100%\\" id=\\"kittens\\" class=\\"quicktube__iframe\\"></iframe></div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster quicktube__poster--hidden\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube YT Player events onStateChange PAUSED 2`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "",
  "Video Paused - test title",
]
`;

exports[`Quicktube YT Player events onStateChange PLAYING 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube quicktube--playing\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    <iframe src=\\"https://www.youtube.com/embed/k6QanQUaDOo?autoplay=1&amp;showinfo=0&amp;autohide=1&amp;color=white&amp;enablejsapi=1&amp;playerapiid=ytplayer&amp;wmode=transparent\\" width=\\"100%\\" id=\\"kittens\\" class=\\"quicktube__iframe\\"></iframe></div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster quicktube__poster--hidden\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube YT Player events onStateChange PLAYING 2`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "",
  "Video Played - test title",
]
`;

exports[`Quicktube YT Player events onStateChange PLAYING monitoring still PLAYING 0% 1`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "",
  "0% Video played - test title",
]
`;

exports[`Quicktube YT Player events onStateChange PLAYING monitoring still PLAYING 25% 1`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "",
  "25% Video played - test title",
]
`;

exports[`Quicktube YT Player events onStateChange PLAYING monitoring still PLAYING 50% 1`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "",
  "50% Video played - test title",
]
`;

exports[`Quicktube YT Player events onStateChange PLAYING monitoring still PLAYING 100% 1`] = `
Array [
  "send",
  "event",
  "Youtube Videos",
  "",
  "50% Video played - test title",
]
`;

exports[`Quicktube loaded 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    </div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube onYouTubeIframeAPIReady initialises on page ready 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    </div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube play click 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube quicktube--playing\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    <iframe src=\\"https://www.youtube.com/embed/k6QanQUaDOo?autoplay=1&amp;showinfo=0&amp;autohide=1&amp;color=white&amp;enablejsapi=1&amp;playerapiid=ytplayer&amp;wmode=transparent\\" width=\\"100%\\" id=\\"kittens\\" class=\\"quicktube__iframe\\"></iframe></div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster quicktube__poster--hidden\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube play iOS click 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\"><iframe src=\\"https://www.youtube.com/embed/k6QanQUaDOo?autoplay=1&amp;showinfo=0&amp;autohide=1&amp;color=white&amp;enablejsapi=1&amp;playerapiid=ytplayer&amp;wmode=transparent\\" width=\\"100%\\" id=\\"kittens\\" class=\\"quicktube__iframe\\"></iframe></div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
"
`;

exports[`Quicktube play keydown right key 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube quicktube--playing\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    <iframe src=\\"https://www.youtube.com/embed/k6QanQUaDOo?autoplay=1&amp;showinfo=0&amp;autohide=1&amp;color=white&amp;enablejsapi=1&amp;playerapiid=ytplayer&amp;wmode=transparent\\" width=\\"100%\\" id=\\"kittens\\" class=\\"quicktube__iframe\\"></iframe></div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster quicktube__poster--hidden\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube play keydown wrong key 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    </div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;

exports[`Quicktube stop click 1`] = `
"
<script src=\\"https://www.youtube.com/iframe_api\\"></script><script></script>
<div class=\\"quicktube\\" data-quicktube=\\"kittens\\">
    <div class=\\"quicktube__video quicktube__video--paused\\" data-quicktube-video=\\"k6QanQUaDOo\\">
    </div>
    <div data-quicktube-play=\\"kittens\\" class=\\"quicktube__poster\\" data-quicktube-poster=\\"\\">
        <div class=\\"quicktube__play quicktube__btn\\">
            Play
        </div>
    </div>
</div>
<div data-quicktube-stop=\\"\\">Stop button</div>
"
`;
